### PR TITLE
fix: Ignore emoji shorthand codes in URIs

### PR DESCRIPTION
**Pull request by socsieng**
Fixes: #1823,
Fixes: #1839

## **Summary**

Fixes an [issue](https://github.com/docsifyjs/docsify/issues/1823) related to unintentional processing of emoji shorthand codes when included in a URI

## **What kind of change does this PR introduce?**

Bugfix

## **For any code change,**

- [x] Related documentation has been updated if needed
- [x] Related tests have been updated or tests have been added

## **Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

## **Tested in the following browsers:**

Using: https://codesandbox.io/s/thirsty-fog-inwfl8

- [x] Chrome
- [ ] Firefox
- [x] Safari
- [ ] Edge
- [ ] IE

### CODE REVIEW

File: src/core/render/emojify.js [hunk 1]
```diff
@@ -35,6 +35,8 @@ export function emojify(text, useNativeEmoji) {
       )
       // Mark colons in comments
       .replace(/<!--[\s\S]+?-->/g, m => m.replace(/:/g, '__colon__'))
+      // Mark colons in URIs
+      .replace(/[a-z]{2,}:\/\/[^\s]+/gi, m => m.replace(/:/g, '__colon__'))
```
Comment (jhildenbiddle) [inline src/core/render/emojify.js:39]: This regex will miss URLs that omit the protocol:

```html
<a href="//domain.com/:smile/">Smile</a>
```

**Review by jhildenbiddle — CHANGES_REQUESTED** [review]
Thanks for submitting the PR, @socsieng.

There is a much easier fix: simply add `a` to the existing list of tags that are being marked:

https://github.com/docsifyjs/docsify/blob/682bf967aed9bc557c688360d9ca61b81b4d2db0/src/core/render/emojify.js#L31-L35

Change this line:

https://github.com/docsifyjs/docsify/blob/682bf967aed9bc557c688360d9ca61b81b4d2db0/src/core/render/emojify.js#L33

To this:

```js
/<(a|code|pre|script|template)[^>]*?>[\s\S]+?<\/(a|code|pre|script|template)>/g,
```

This fix also addresses an issue with the regex you were using in your initial PR.

**Comment by socsieng** [comment]
Wouldn't marking the `a` tags be too broad? I can imagine emojis being used in many `a` tags today.

> This fix also addresses an issue with the regex you were using in your initial PR.

What was the issue with the regex?

File: src/core/render/emojify.js (hunk 1 above)
Comment (socsieng) [reply src/core/render/emojify.js:39]: I was trying to be a strict as possible with the regex to reduce the number of false positives. Happy to update the expression to make `scheme:` optional.

The root cause of my issue was related to absolute SharePoint urls being pasted into markdown which were all absolute and included the scheme.

Comment (socsieng) [reply src/core/render/emojify.js:39]: Note that //domain.com/:smile:/ doesn't https://domain.com/:smile:/ appear to be automatically parsed as a URI like a fully qualified URI does by the markdown parser.

What do you prefer? Optional versus required `scheme:`?

Comment (socsieng) [inline src/core/render/emojify.js:39]: Can change to the following if you would like it to be optional:

```suggestion
      .replace(/([a-z]{2,}:)?\/\/[^\s]+/gi, m => m.replace(/:/g, '__colon__'))
```

Comment (jhildenbiddle) [reply src/core/render/emojify.js:39]: You're right about ignoring everything inside `<a>` tags. That will miss some shortcodes that I believe people will expect to be processed.

There are a number of scenarios to consider. We should probably start by defining the behavior. For example:

```html
<style>
  @import url("http://domain.com/:smile/file.css");
  @import url('http://domain.com/:smile/file.css');
  @import url(http://domain.com/:smile/file.css);
  @import "http://domain.com/:smile/file.css";
  @import 'http://domain.com/:smile/file.css';

  .myclass::before {
    content: ":smile:";
    background: url("http://domain.com/:smile/");
    background: url('http://domain.com/:smile/');
    background: url(http://domain.com/:smile/);
  }
</style>

<a href="http://domain.com/:smile:/">:smile: http://domain.com/:smile:/ :smile:</a>
<p>:smile: http://domain.com/:smile:/ :smile:</p>

<img src="http://domain.com/:smile:/file.png" alt=":smile:">
<img srcset="http://domain.com/:smile:/file.jpg 320w" src="http://domain.com/:smile:/file.png">

<script src="http://domain.com/:smile:/file.js"></script>
```

Here's how I would assume we would want to final output to look like:

```html
<style>
  @import url("http://domain.com/:smile/file.css");
  @import url('http://domain.com/:smile/file.css');
  @import url(http://domain.com/:smile/file.css);
  @import "http://domain.com/:smile/file.css";
  @import 'http://domain.com/:smile/file.css';

  .myclass::before {
    content: "😄";
    background: url("http://domain.com/:smile/");
    background: url('http://domain.com/:smile/');
    background: url(http://domain.com/:smile/);
  }
</style>

<a href="http://domain.com/:smile:/">😄 http://domain.com/:smile:/ 😄</a>
<p>😄 http://domain.com/:smile:/ 😄</p>

<img src="http://domain.com/:smile:/file.png" title="😄">
<img srcset="http://domain.com/:smile:/file.jpg 320w" src="http://domain.com/:smile:/file.png">

<script src="http://domain.com/:smile:/file.js"></script>
```

I'm probably missing something, but the rules for the above scenarios would be:

1. Ignore emoji shortcodes in `href`, `src`, and `srcset` attributes
1. Ignore emoji shortcodes in `url()` values
1. Ignore emoji shortcodes in text URLs within tags

Thoughts?

Comment (socsieng) [reply src/core/render/emojify.js:39]: I think that looks about right. 3 is a useful _catch all_ for URLs anywhere in the content, while 1 and 2 are useful for handling partial URLs (e.g. relative).

For my use case, 3 is sufficient and what this pull request is meant to fix. 1 and 2 are nice to have.

Comment (jhildenbiddle) [reply src/core/render/emojify.js:39]: I believe we can handle all three scenarios by making a small change to your suggested change above:

```js
.replace(/([a-z]{2,}:)?\/\/[^\s'">)]+/gi, m => m.replace(/:/g, '__colon__'))
```

Tested here: https://regex101.com/r/Paa6KG/1

**Review by jhildenbiddle — CHANGES_REQUESTED** [review]

File: src/core/render/emojify.js (hunk 1 above)
Comment (socsieng) [reply src/core/render/emojify.js]: Fixed

Comment (socsieng) [reply src/core/render/emojify.js]: Fixed

**Review by jhildenbiddle — APPROVED** [review]
LGTM! Thanks again, @socsieng.

**Review by sy-records — APPROVED** [review]
LGTM. Thank you!

**Comment by socsieng** [comment]
Any ideas about when these changes will be available in npm?

**Comment by sy-records** [comment]
Not sure when npm will be released.

But you can try https://github.com/sy-records/docsify-nightly

**Comment by trusktr** [comment]
You can also try installing from github using `npm`. To install from `develop`:

```sh
npm install "docsify@docsifyjs/docsify#develop"
```

Or to install from a commit:


```sh
npm install "docsify@docsifyjs/docsify#3c9b3d9702bb05a5ff45a4ce4233e144cf1ebecb"
```

Or any git ref:


```sh
npm install "docsify@docsifyjs/docsify#<GIT_REF_GOES_HERE>"
```

which can be a tag, branch, etc.

**Comment by socsieng** [comment]
Thanks. Appreciate the help.

I was more wondering what the timelines were to determine if I should wait for a formal package update or if I should use a github based version.

**Comment by jhildenbiddle** [comment]
@socsieng --

I think we're close to pushing an update. There's one other PR that I'd like to see merged before we do so, but otherwise I don't think anyone would have an issue pushing a new minor or patch release (have to check history to determine which would be appropriate). I'll see if I can wrap that up today or tomorrow.

**Pull request merged**